File: packages/core/src/server/routes/tags.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import type { RouteContext } from './types.js';
import { param } from './types.js';
import { TAG_PALETTE } from '@qlan-ro/mainframe-types';
import { createChildLogger } from '../../logger.js';

const logger = createChildLogger('routes:tags');

const ColorSchema = z.enum(TAG_PALETTE);
const CreateBody = z.object({ name: z.string(), color: ColorSchema.optional() });
const PatchBody = z
  .object({ rename: z.string().optional(), color: ColorSchema.optional() })
  .refine((v) => v.rename !== undefined || v.color !== undefined, {
    message: 'rename or color required',
  });
const SetChatTagsBody = z.object({ tags: z.array(z.string()) });

export function tagRoutes(ctx: RouteContext): Router {
  const router = Router();

  router.get('/api/tags', (_req: Request, res: Response) => {
    res.json({ success: true, data: ctx.db.tags.list() });
  });

  router.post('/api/tags', (req: Request, res: Response) => {
    const parsed = CreateBody.safeParse(req.body);
    if (!parsed.success) {
      res.status(400).json({ success: false, error: parsed.error.message });
      return;
    }
    try {
      const tag = ctx.db.tags.upsert(parsed.data.name, parsed.data.color);
      res.status(201).json({ success: true, data: tag });
    } catch (err) {
      logger.warn({ err }, 'create tag failed');
      res.status(400).json({ success: false, error: String((err as Error).message) });
    }
  });

  router.patch('/api/tags/:name', (req: Request, res: Response) => {
    const parsed = PatchBody.safeParse(req.body);
    if (!parsed.success) {
      res.status(400).json({ success: false, error: parsed.error.message });
      return;
    }
    const name = param(req, 'name');
    if (!ctx.db.tags.get(name)) {
      res.status(404).json({ success: false, error: 'Tag not found' });
      return;
    }
    try {
      if (parsed.data.rename !== undefined) ctx.db.tags.rename(name, parsed.data.rename);
      const final = parsed.data.rename ?? name;
      if (parsed.data.color !== undefined) ctx.db.tags.setColor(final, parsed.data.color);
      const result = ctx.db.tags.get(final);
      res.json({ success: true, data: result });
    } catch (err) {
      logger.warn({ err, name }, 'update tag failed');
      res.status(400).json({ success: false, error: String((err as Error).message) });
    }
  });

  router.delete('/api/tags/:name', (req: Request, res: Response) => {
    const name = param(req, 'name');
    if (!ctx.db.tags.get(name)) {
      res.status(404).json({ success: false, error: 'Tag not found' });
      return;
    }
    ctx.db.tags.remove(name);
    res.status(204).end();
  });

  router.get('/api/chats/:id/tags', (req: Request, res: Response) => {
    const tags = ctx.db.chatTags.listForChat(param(req, 'id'));
    res.json({ success: true, data: tags });
  });

  router.put('/api/chats/:id/tags', (req: Request, res: Response) => {
    const parsed = SetChatTagsBody.safeParse(req.body);
    if (!parsed.success) {
      res.status(400).json({ success: false, error: parsed.error.message });
      return;
    }
    const chatId = param(req, 'id');
    try {
      ctx.db.chatTags.setForChat(chatId, parsed.data.tags, ctx.db.tags);
      const persisted = ctx.db.chatTags.listForChat(chatId);
      ctx.chats.syncChatTags(chatId, persisted);
      res.json({ success: true, data: persisted });
    } catch (err) {
      logger.warn({ err, chatId }, 'set chat tags failed');
      res.status(400).json({ success: false, error: String((err as Error).message) });
    }
  });

  return router;
}

File: packages/core-rs/crates/mainframe-server/src/routes/tags.rs
//! Ported from `src/server/routes/tags.ts` — tag CRUD + per-chat tag assignment.
//!
//! `TagColor` is the shared enum (`z.enum(TAG_PALETTE)` → serde parse; an
//! out-of-palette color fails the body parse → 400). Mutating handlers wrap the
//! repo call and surface any error as a 400 with the verbatim message (the TS
//! `try/catch → String(err.message)`); the read handlers let an error fall
//! through to the opaque 500 (the TS un-wrapped path). DELETE deviates from the
//! rest of the API on purpose: it ends 204 with NO body, not the `{success}`
//! envelope.

use std::sync::Arc;

use axum::Router;
use axum::body::Bytes;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::routing::{get, patch};
use mainframe_types::tags::TagColor;
use serde::Deserialize;

use crate::ctx::AppCtx;
use crate::respond::{fail, ok};
use crate::routes::projects::parse_body;

#[derive(Deserialize)]
struct CreateBody {
    name: String,
    color: Option<TagColor>,
}

#[derive(Deserialize)]
struct PatchBody {
    rename: Option<String>,
    color: Option<TagColor>,
}

#[derive(Deserialize)]
struct SetChatTagsBody {
    tags: Vec<String>,
}

fn bad_request(err: &mainframe_db::DbError) -> Response {
    fail(StatusCode::BAD_REQUEST, err.to_string())
}

async fn list(State(ctx): State<Arc<AppCtx>>) -> Response {
    match ctx.db.call(|db| db.tags.list()).await {
        Ok(tags) => ok(tags),
        Err(err) => crate::async_err::internal_error("list tags", &err),
    }
}

async fn create(State(ctx): State<Arc<AppCtx>>, body: Bytes) -> Response {
    let Some(parsed): Option<CreateBody> = parse_body(&body) else {
        return fail(StatusCode::BAD_REQUEST, "Invalid request body");
    };
    let CreateBody { name, color } = parsed;
    match ctx.db.call(move |db| db.tags.upsert(&name, color)).await {
        Ok(tag) => (StatusCode::CREATED, ok(tag)).into_response(),
        Err(err) => bad_request(&err),
    }
}

async fn update(State(ctx): State<Arc<AppCtx>>, Path(name): Path<String>, body: Bytes) -> Response {
    let Some(parsed): Option<PatchBody> = parse_body(&body) else {
        return fail(StatusCode::BAD_REQUEST, "Invalid request body");
    };
    if parsed.rename.is_none() && parsed.color.is_none() {
        return fail(StatusCode::BAD_REQUEST, "rename or color required");
    }

    let lookup = name.clone();
    match ctx.db.call(move |db| db.tags.get(&lookup)).await {
        Ok(Some(_)) => {}
        Ok(None) => return fail(StatusCode::NOT_FOUND, "Tag not found"),
        Err(err) => return crate::async_err::internal_error("get tag", &err),
    }

    let PatchBody { rename, color } = parsed;
    let result = ctx
        .db
        .call(move |db| {
            if let Some(ref new_name) = rename {
                db.tags.rename(&name, new_name)?;
            }
            let final_name = rename.as_deref().unwrap_or(&name).to_string();
            if let Some(color) = color {
                db.tags.set_color(&final_name, color)?;
            }
            db.tags.get(&final_name)
        })
        .await;
    match result {
        Ok(tag) => ok(tag),
        Err(err) => bad_request(&err),
    }
}

async fn remove(State(ctx): State<Arc<AppCtx>>, Path(name): Path<String>) -> Response {
    let lookup = name.clone();
    match ctx.db.call(move |db| db.tags.get(&lookup)).await {
        Ok(Some(_)) => {}
        Ok(None) => return fail(StatusCode::NOT_FOUND, "Tag not found"),
        Err(err) => return crate::async_err::internal_error("get tag", &err),
    }
    match ctx.db.call(move |db| db.tags.remove(&name)).await {
        Ok(()) => StatusCode::NO_CONTENT.into_response(),
        Err(err) => crate::async_err::internal_error("remove tag", &err),
    }
}

async fn list_chat_tags(State(ctx): State<Arc<AppCtx>>, Path(id): Path<String>) -> Response {
    match ctx.db.call(move |db| db.chat_tags.list_for_chat(&id)).await {
        Ok(tags) => ok(tags),
        Err(err) => crate::async_err::internal_error("list chat tags", &err),
    }
}

async fn set_chat_tags(
    State(ctx): State<Arc<AppCtx>>,
    Path(id): Path<String>,
    body: Bytes,
) -> Response {
    let Some(parsed): Option<SetChatTagsBody> = parse_body(&body) else {
        return fail(StatusCode::BAD_REQUEST, "Invalid request body");
    };
    let tags = parsed.tags;
    let result = ctx
        .db
        .call(move |db| {
            db.chat_tags.set_for_chat(&id, &tags, &db.tags)?;
            db.chat_tags.list_for_chat(&id)
        })
        .await;
    match result {
        // ctx.chats?.syncChatTags is an optional Phase-4/5 ChatManager hook — a
        // no-op when the manager is absent, so it is intentionally omitted here.
        Ok(persisted) => ok(persisted),
        Err(err) => bad_request(&err),
    }
}

pub fn router() -> Router<Arc<AppCtx>> {
    Router::new()
        .route("/api/tags", get(list).post(create))
        .route("/api/tags/{name}", patch(update).delete(remove))
        .route(
            "/api/chats/{id}/tags",
            get(list_chat_tags).put(set_chat_tags),
        )
}

// PORT STATUS: src/server/routes/tags.ts (6 endpoints, 98 lines)
// confidence: high
// todos: 0
// notes: z.enum(TAG_PALETTE) → Option<TagColor> serde parse (out-of-palette →
// body-parse fail → 400). PatchBody's `.refine(rename||color)` → explicit
// both-None → 400 "rename or color required". Mutating handlers map DbError →
// 400 with the verbatim message (TS try/catch → String(err.message), which also
// carries the validate-tag-name reserved/short strings); read handlers → opaque
// 500. rename+setColor+get run in ONE db.call so the read-back is atomic on the
// DB thread. DELETE ends 204 with an empty body (the pinned deviation). The
// optional ctx.chats.syncChatTags hook is a Phase-4/5 no-op and is omitted.
